### PR TITLE
go.mod: update to github.com/emicklei/go-restful/v3 v3.7.3

### PR DESCRIPTION
**Pull request by thaJeztah**
full diff: https://github.com/emicklei/go-restful/compare/v2.9.5...v3.7.3

- Switch to using go modules
- Add check for wildcard to fix CORS filter
- Add check on writer to prevent compression of response twice
- Add OPTIONS shortcut WebService receiver
- Add Route metadata to request attributes or allow adding attributes to routes
- Add wroteHeader set
- Enable content encoding on Handle and ServeHTTP
- Feat: support google custom verb
- Feature: override list of method allowed without content-type
- Fix Allow header not set on '405: Method Not Allowed' responses
- Fix Go 1.15: conversion from int to string yields a string of one rune
- Fix WriteError return value
- Fix: use request/response resulting from filter chain
- handle path params with prefixes and suffixes
- HTTP response body was broken, if struct to be converted to JSON has boolean value
- List available representations in 406 body
- Support describing response headers
- Unwrap function in filter chain + remove unused dispatchWithFilters

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Updating this dependency to a version that's using go modules

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

The non-go module version is still in use by various other dependencies within the k8s.io org, so those will need to be updated as well (and updated in this repository after they've been updated)

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -229,6 +229,7 @@ replace (
 	github.com/dustin/go-humanize => github.com/dustin/go-humanize v1.0.0
 	github.com/elazarl/goproxy => github.com/elazarl/goproxy v0.0.0-20180725130230-947c36da3153 // 947c36da3153 is the SHA for git tag v1.11
 	github.com/emicklei/go-restful => github.com/emicklei/go-restful v2.9.5+incompatible
+	github.com/emicklei/go-restful/v3 => github.com/emicklei/go-restful/v3 v3.7.3
```
Comment (thaJeztah) [inline go.mod:232]: I've left the `github.com/emicklei/go-restful` (non-go modules) replace in place; I wasn't sure what's common in this scenario, but that replace should (if I'm correct) still help with getting the indirect dependencies to the same / consistent version.

(@dims ?)

**Comment by thaJeztah** [comment]
On this branch; indirect dependencies remaining that use the old (non-go modules) version;

```
go mod graph | grep ' github.com/emicklei/go-restful@'
k8s.io/apiserver@v0.0.0 github.com/emicklei/go-restful@v2.9.5+incompatible
k8s.io/code-generator@v0.0.0 github.com/emicklei/go-restful@v2.9.5+incompatible
k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65 github.com/emicklei/go-restful@v0.0.0-20170410110728-ff4f55a20633
k8s.io/kube-openapi@v0.0.0-20210421082810-95288971da7e github.com/emicklei/go-restful@v0.0.0-20170410110728-ff4f55a20633
```

Perhaps a replace rule is needed for `k8s.io/kube-openapi`, as it looks like multiple versions are considered by go modules

**Comment by thaJeztah** [comment]
Opened https://github.com/kubernetes/kube-openapi/pull/271 to open kube-openapi

**Comment by liggitt** [comment]
thanks for the PR... will need to sweep the go-restful release notes and impl changes from v2 to v3 to see if any behavior changes impact us

xref comments on previous bump attempt:
* https://github.com/kubernetes/kubernetes/pull/97961#discussion_r557004785
* https://github.com/kubernetes/kubernetes/pull/97961#issuecomment-775325602
* https://github.com/kubernetes/kubernetes/pull/97961#issuecomment-871759827

is there a specific bug/issue this bump is intended to address? that would help in knowing how to prioritize review

**Comment by thaJeztah** [comment]
Ouch; I was afraid this could happen; guess either some code needs to be regenerated, or we need to update this in a specific order 😞 

```
# k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/routes
vendor/k8s.io/apiserver/pkg/server/routes/openapi.go:39:64: cannot use c.RegisteredWebServices() (type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService) as type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful".WebService in argument to builder.BuildOpenAPISpec

vendor/k8s.io/apiserver/pkg/server/routes/openapi.go:78:41: cannot use ws (type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService) as type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful".WebService in argument to builder3.BuildOpenAPISpec
# k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder
vendor/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:203:56: cannot use []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService{...} (type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService) as type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful".WebService in argument to builder3.BuildOpenAPISpec
vendor/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:213:62: cannot use []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService{...} (type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".WebService) as type []*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful".WebService in argument to builder.BuildOpenAPISpec
vendor/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:529:3: cannot use "k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/openapi".GetOperationIDAndTags (type func(*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/v3".Route) (string, []string, error)) as type func(*"k8s.io/kubernetes/vendor/github.com/emicklei/go-restful".Route) (string, []string, error) in field value
```

**Comment by thaJeztah** [comment]
> is there a specific bug/issue this bump is intended to address? that would help in knowing how to prioritize review

No, there's no specific bug/issue; mostly that I noticed the go-restful project moved to using go modules, and I noticed that containerd was still using the non-go modules version (then I noticed there were various indirect dependencies using it).

Based on the errors above, it looks like updating may be a bit hairy as various components use the type 😬 (I'm not very familiar with the dependency tree / relation here, but happy to help getting changes in, in the right order)

**Comment by liggitt** [comment]
superceded by https://github.com/kubernetes/kubernetes/pull/110351

/close

**Pull request closed without merging**